File: src/test/java/ru/romanow/jpa/repository/PersonRepositoryTest.java
package ru.romanow.jpa.repository;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import ru.romanow.jpa.domain.Authority;
import ru.romanow.jpa.domain.Person;

import javax.persistence.EntityManager;
import java.util.Objects;
import java.util.Set;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;


@DataJpaTest()
class PersonRepositoryTest {

    @Autowired
    private EntityManager em;

    @Test
    void shouldSavePersonWithAllAuthority() {
        final var savedPerson = makeSavedPersonWithAuthorities();

        assertThat(savedPerson).isNotNull();
        assertThat(savedPerson.getAuthorities()).isNotNull();
        assertThat(savedPerson.getAuthorities()).allMatch(Objects::nonNull);
    }


    @Test
    void shouldDeleteAllAuthoritiesWithPerson() {
        final var savedPerson = makeSavedPersonWithAuthorities();
        assertThat(getAuthorityCount()).isNotZero();

        em.remove(savedPerson);

        em.flush();
        em.clear();

        assertThat(getAuthorityCount()).isZero();
    }

    private Long getAuthorityCount() {
        return (Long) em.createQuery("select count(a) from Authority a").getSingleResult();
    }

    private Person makeSavedPersonWithAuthorities() {
        final var authority1 = new Authority()
                .setName("authority1")
                .setPriority(1);
        final var authority2 = new Authority()
                .setName("authority2")
                .setPriority(2);
        final var person = new Person()
                .setFirstName("firstName")
                .setLastName("lastName")
                .setAuthorities(Set.of(authority1, authority2));
        em.persist(person);
        em.flush();
        em.clear();
        return em.find(Person.class, person.getId());
    }
}

File: src/test/java/ru/romanow/jpa/repository/PersonRepositoryTest.kt
package ru.romanow.jpa.repository

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import ru.romanow.jpa.config.DatabaseTestConfiguration
import ru.romanow.jpa.utils.AUTHORITY_COUNT
import ru.romanow.jpa.utils.ROLE_COUNT
import ru.romanow.jpa.utils.buildPerson

@DataJpaTest
@Import(DatabaseTestConfiguration::class)
internal class PersonRepositoryTest {

    @Autowired
    private lateinit var personRepository: PersonRepository

    @Autowired
    private lateinit var authorityRepository: AuthorityRepository

    @Test
    fun `when create new Person then save Authorities too`() {
        val person = buildPerson(ROLE_COUNT, AUTHORITY_COUNT)
        val authorities = person.authorities.map { it.name }.toTypedArray()
        val saved = personRepository.save(person)

        assertThat(saved).isNotNull()
        assertThat(saved.authorities).hasSize(3)
        assertThat(saved.authorities)
            .extracting("name")
            .containsExactlyInAnyOrder(*authorities)
    }

    @Test
    fun `when delete Person then delete Authorities too`() {
        val person = personRepository.save(buildPerson(ROLE_COUNT, AUTHORITY_COUNT))
        assertThat(authorityRepository.count()).isNotZero()

        personRepository.delete(person)
        personRepository.flush()

        assertThat(authorityRepository.count()).isZero()
    }
}
